Two Factor: Allow users to log in with backup codes and improve the security keys flow

xknown opened the pull request:
Users with a TOTP based second factor (either SMS or an App) are able to enter backup codes directly into the verification form when they log in, however it's not clear for most users they can use them.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

## Proposed Changes

* We add an explicit button "Continue with a backup code" to instruct users to enter a 2fa backup code instead.
Before
<img width="404" alt="Screenshot 2024-02-06 at 12 03 09" src="https://github.com/Automattic/wp-calypso/assets/1081870/52b1e081-4b56-4cc4-b560-d22f2ae1648d">
Verification form
<img width="410" alt="Screenshot 2024-02-06 at 12 03 17" src="https://github.com/Automattic/wp-calypso/assets/1081870/378cea04-d48b-4f31-aaf5-729a212719c7">

After
<img width="386" alt="Screenshot 2024-02-06 at 12 02 39" src="https://github.com/Automattic/wp-calypso/assets/1081870/455dc621-cca7-462b-a370-eecaf3cc96da">
Verification form for backup codes
<img width="406" alt="Screenshot 2024-02-06 at 12 02 51" src="https://github.com/Automattic/wp-calypso/assets/1081870/0edd75f6-ed22-4f1d-accf-5b4baf0b24f9">


* We also remove the requirement to click on the "Continue with security key" button when signing in with a security key. If it fails for some reason, the old button will still work as always if the user decides to retry the same action.


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Open an incognito window.
* Log in with a WP.com user using a security key as second factor (you might need to add a specific key for the calypso development environment).
* Notice you don't need to click any button before the security key flow starts.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

xknown committed:
Two Factor Security Keys: Automatically prompt for a security key.

The current 2fa authentication flow with security keys require users to
click on the "Continue with security key" button before we attempting to
initiate this flow.

This change automatically asks for a security key if the given account
supports it. If it fails for some reason, the old button will still work
as always if the user decides to retry the same action.

xknown committed:
Allow users to log in with backup codes when the enhanced security mode
is on.

tyxla commented on client/blocks/login/two-factor-authentication/security-key-form.jsx:
```diff
@@ -31,8 +31,12 @@ class SecurityKeyForm extends Component {
 	};
 
+	componentDidMount() {
+		setTimeout( () => this.initiateSecurityKeyAuthentication(), 100 );
```
Why is the timeout necessary? 🤔 
Also, why 100ms?

xknown replied:
Removed the `setTimeout` statement in 1fb7185df4, I initially added it because I thought for some reason it was "needed".

tyxla commented on client/blocks/login/two-factor-authentication/security-key-form.jsx:
```diff
@@ -31,8 +31,12 @@ class SecurityKeyForm extends Component {
+
 	initiateSecurityKeyAuthentication = ( event ) => {
-		event.preventDefault();
+		event?.preventDefault();
```
This feels like a code smell, and a sign that we'll have to separate the event handling from what happens afterwards in 2 separate methods.

xknown replied:
Removed the `event.preventDefault()` call in 1fb7185df4 which as you point out does seem to be needed.

xknown replied:
Reverted the removal of `event.preventDefault` in [21517fd](https://github.com/Automattic/wp-calypso/pull/87128/commits/21517fdf68b3183f8f877b2df2cc51c7bd04e09c). This function is used on a form's `onSubmit` event, without it the form is submitted which causes a full page reload.

tyxla commented on client/blocks/login/two-factor-authentication/two-factor-actions.jsx:
```diff
@@ -89,6 +105,12 @@ class TwoFactorActions extends Component {
 					) }
+
+					{ isBackupCodeAvailable && (
+						<Button data-e2e-link="2fa-otp-link" onClick={ this.recordBackupLinkClick }>
```
Are we going to specifically need the `data-e2e-link` data attribute? If that's not certain, let's omit it for now.

xknown replied:
Not really, removed it in 17fd7c3c25

tyxla commented on client/blocks/login/two-factor-authentication/two-factor-actions.jsx:
```diff
@@ -44,6 +44,15 @@ class TwoFactorActions extends Component {
 	};
+
+	recordBackupLinkClick = ( event ) => {
+		event.preventDefault();
```
I see this handler is attached to the `onClick` event of a button without a `href`. Is `event.preventDefault()` necessary then?

tyxla commented on client/blocks/login/two-factor-authentication/two-factor-actions.jsx:
```diff
@@ -44,6 +44,15 @@ class TwoFactorActions extends Component {
+	recordBackupLinkClick = ( event ) => {
+		event.preventDefault();
+
+		this.props.recordTracksEvent( 'calypso_login_two_factor_switch_to_backup_link_click' );
```
Oh, that's a long event name. But it seems to meet the guidelines (PCYsg-4S2-p2) and fits the rest of the events here, so maybe lets keep it as-is.

xknown committed:
Remove useless code.

This removes the `setTimeout()` call as well as the `event.preventDefault()`
statements which don't seem to be needed.

xknown committed:
Remove the data-e2e-link attribute

xknown committed:
Remove useless setTimeout/event.preventDefault statements.

xknown committed:
Reinstante event.preventDefault()

This partially reverts 1fb7185df46f96416b311101809ab4f56a7ec7f4 and
31663daaa38d45465e3912519cc4f16860ef3366

`event.preventDefault()` is needed one one manually clicks on the
"Continue with security" key button after a failed authentication
attempt. If it's missing, the form is submitted and there's a full page
reload.

tyxla commented:
@xknown thanks for the updates and for addressing the feedback! Let me know if you're still working on it or if it's ready for another look.

tyxla reviewed (approved):
This looks good to me 👍 🚀
